internal/era/era.go

```diff
@@ -0,0 +1,336 @@
+}
+
+// Read reads one (block, receipts) tuple from an Era1 archive.
+func (r *Reader) Read() (*types.Block, types.Receipts, error) {
+	block, receipts, err := r.ReadBlockAndReceipts(r.next)
+	if err != nil {
+		return nil, nil, err
+	}
+	r.next += 1
+	return block, receipts, nil
+}
+
+// readBlob reads an entry of data.
+func (r *Reader) readEntry(n uint64, skip int) (*e2store.Entry, error) {
+	if n < r.metadata.start || r.metadata.start+r.metadata.count < n {
+		return nil, fmt.Errorf("request out-of-bounds: want %d, start: %d, count: %d", n, r.metadata.start, r.metadata.count)
+	}
+	// Read the specified block's offset from the block index.
+	off, err := r.readOffset(n)
+	if err != nil {
+		return nil, fmt.Errorf("error reading block offset: %w", err)
+	}
+	// Skip to the requested entry.
+	for i := 0; i < skip; i++ {
+		if length, err := r.e.LengthAt(off); err != nil {
+			return nil, err
+		} else {
+			off += length
+		}
+	}
+	// Read entry.
+	var entry e2store.Entry
+	if _, err := r.e.ReadAt(&entry, off); err != nil {
+		return nil, err
+	}
+	return &entry, nil
+}
+
+// readHeaderRLP reads the header number n RLP.
+func (r *Reader) readHeaderRLP(n uint64) ([]byte, error) {
+	e, err := r.readEntry(n, 0)
+	if err != nil {
+		return nil, err
+	}
+	if e.Type != TypeCompressedHeader {
+		return nil, fmt.Errorf("expected header entry, got %x", e.Type)
+	}
+	return io.ReadAll(snappy.NewReader(bytes.NewReader(e.Value)))
+}
+
+// readBodyRLP reads the block body number n RLP.
+func (r *Reader) readBodyRLP(n uint64) ([]byte, error) {
+	e, err := r.readEntry(n, 1)
+	if err != nil {
+		return nil, err
+	}
+	if e.Type != TypeCompressedBody {
+		return nil, fmt.Errorf("expected body entry, got %x", e.Type)
+	}
+	return io.ReadAll(snappy.NewReader(bytes.NewReader(e.Value)))
+}
+
+// readReceiptsRLP reads the receipts RLP associated with number n.
+func (r *Reader) readReceiptsRLP(n uint64) ([]byte, error) {
+	e, err := r.readEntry(n, 2)
+	if err != nil {
+		return nil, err
+	}
+	if e.Type != TypeCompressedReceipts {
+		return nil, fmt.Errorf("expected receipts entry, got %x", e.Type)
+	}
+	return io.ReadAll(snappy.NewReader(bytes.NewReader(e.Value)))
+}
+
+// readTotalDifficulty reads the total difficulty of block number n.
+func (r *Reader) readTotalDifficulty(n uint64) (*big.Int, error) {
+	e, err := r.readEntry(n, 3)
+	if err != nil {
+		return nil, err
+	}
+	if e.Type != TypeTotalDifficulty {
+		return nil, fmt.Errorf("expected TD entry, got %x", e.Type)
+	}
+	return new(big.Int).SetBytes(reverseOrder(e.Value)), nil
+}
+
+// ReadHeader reads the header number n.
+func (r *Reader) ReadHeader(n uint64) (*types.Header, error) {
+	h, err := r.readHeaderRLP(n)
+	if err != nil {
+		return nil, err
+	}
+	var header types.Header
+	if err := rlp.DecodeBytes(h, &header); err != nil {
+		return nil, err
+	}
+	return &header, nil
+}
+
+// ReadBlock reads the block number n.
+func (r *Reader) ReadBlock(n uint64) (*types.Block, error) {
```

Comment:
One thing that this implementation does quite a bit, is handle full arrays. This is often the simplest way to deal with data, but leads to large overhead in processing. 

Like, a file is read in full from disk. It is marshalled and copied into a slice in a struct. Then that data is decompressed into a new buffer. Then that buffer given to `rlp` to marshal into a `block` struct. 

A lot of memory copying/allocation can be avoided if e.g. `io.Reader`s are used in the pipeline. Here's one example where an `io.Reader` could be used instead: 

```golang
func (r *Reader) At(n uint64, skip int) (io.Reader, error) {
	e, err := r.readEntry(n, skip)
	if err != nil {
		return nil, err
	}
	return snappy.NewReader(bytes.NewReader(e.Value)), nil
}

// ReadBlock reads the block number n.
func (r *Reader) ReadBlock(n uint64) (*types.Block, error) {
	header, err := r.ReadHeader(n)
	if err != nil {
		return nil, err
	}
	var body types.Body
	input, err := r.At(n, 1)
	if err != nil {
		return nil, err
	}
	if rlp.Decode(input, &body); err != nil {
		return nil, err
	}
	return types.NewBlockWithHeader(header).WithBody(body.Transactions, body.Uncles), nil
}
```
The `snappy.Reader` is passed directly to `rlp.Decode`, avoiding the `io.ReadAll` and managing the buffer explicitly. 

I didn't commit this code, because it might conflict with other potential changes that you might want to do in order to shave down on the number of reads (as I commented on yesterday). Just posting this as a potential direction to take things into.

Reply:
something like this `e2store/e2store.go` maybe ? 
```golang
// ReadAt reads one Entry from r at the specified offset.
func (r *Reader) ReaderAt(expectedType uint16, off int64) (io.Reader, error) {
	typ, length, err := r.ReadMetadataAt(off)
	if err != nil {
		return nil, err
	}
	if typ != expectedType {
		return nil, fmt.Errorf("wrong type, want %d have %d", expectedType, typ)
	}
	if length > valueSizeLimit {
		return nil, fmt.Errorf("item larger than item size limit %d: have %d", valueSizeLimit, length)
	}
	return io.NewSectionReader(r.r, off+headerSize, int64(length)), nil
}
```

Reply:
So by using a pipeline like this, the rlp marshaller would 'pull' from the snappy decoder, which would 'pull' from the `e2store` reader, which would 'pull' from the file. And you'd have a pull-based input flow, instead of passing around slices of bytes.

Reply:
Really good insight - thanks for this. I've added the recommend function in c1a1f93a4c900318df6da88060703a5d67af2cbb.